sqlitecluster/SQLiteNode.cpp

```diff
@@ -165,143 +160,69 @@ SQLiteNode::~SQLiteNode() {
-                        uniqueContraintsError = true;
-                    }
-                    // Now see if we can commit. We wait until *after* prepare because for QUORUM transactions, we
-                    // don't send LEADER the approval for this until inside of `prepare`. This potentially makes us
-                    // wait while holding the commit lock for non-concurrent transactions, but I guess nobody else with
-                    // a commit after us will be able to commit, either.
-                    SINFO("[performance] Waiting on leader to say it has committed transaction " << command.calcU64("NewCount"));
-                    SQLiteSequentialNotifier::RESULT waitResult = _leaderCommitNotifier.waitFor(command.calcU64("NewCount"), true);
-                    SINFO("[performance] Leader reported committing transaction " << command.calcU64("NewCount") << ", committing.");
-                    if (uniqueContraintsError) {
-                        SINFO("Got unique constraints error in replication, restarting.");
-                        --_concurrentReplicateTransactions;
-                        db.rollback();
-                        continue;
-                    } else if (waitResult == SQLiteSequentialNotifier::RESULT::CANCELED) {
-                        SINFO("Replication canceled mid-transaction, stopping.");
-                        --_concurrentReplicateTransactions;
-                        db.rollback();
-                        break;
-                    }
+        }
 
-                    // Leader says it has committed this transaction, so we can too.
-                    ++commitAttemptCount;
-                    result = _handleCommitTransaction(db, peer, command.calcU64("NewCount"), command["NewHash"]);
-                    if (result != SQLITE_OK) {
-                        db.rollback();
-                    }
-                }
-            } catch (const SException& e) {
-                SALERT("Caught exception in replication thread. Assuming this means we want to stop following. Exception: " << e.what());
-                goSearchingOnExit = true;
-                --_concurrentReplicateTransactions;
-                db.rollback();
+        // At this point, we're guaranteed to have a message. Process it and then run again.
+        if (SIEquals(command.methodLine, "BEGIN_TRANSACTION")) {
+            auto start = chrono::steady_clock::now();
+            _handleBeginTransaction(db, peer, command);
+            _handlePrepareTransaction(db, peer, command, dequeueTime);
+            auto duration = chrono::steady_clock::now() - start;
+            SINFO("[performance] Wrote replicate transaction in " << chrono::duration_cast<chrono::microseconds>(duration).count() << "us.");
+        } else if (SIEquals(command.methodLine, "COMMIT_TRANSACTION")) {
+            int result = _handleCommitTransaction(db, peer, command.calcU64("NewCount"), command["NewHash"]);
```

Comment:
I think I'm a little confused by this.  If I'm not mistaken, there's no way to guarantee that we'll receive a sequential set of commits with BEGIN > COMMIT order.

So lets say we have journal entries 1, 2 and 3. We could receive BEGIN_TRANSACTION for 1, 3 and 2, and then receive COMMIT for 3, 1 and 2, right? 

Before, this was not a problem because we would only execute the commit in order, but I'm not understanding how we're keeping that order here if we're adding all of them to the same queue as we receive them in line 1579

Reply:
We always receive them in order, and have always received them in order. They only get out-of-order when we hand them out to separate threads.

Reply:
Hmmm. I see, I thought we could receive them out of order. Does that also apply to BEGIN_TRANSACTION/COMMIT_TRANSACTION? So if we receive BEGIN_TRANSACTION journal entry 1 , we'll receive the commit for that transaction before begin transaction 2?

Reply:
This is generally how they are sent:
https://github.com/Expensify/Bedrock/blob/8d609f3005af4b5baded62e49e2d0747333f68f6/sqlitecluster/SQLiteNode.cpp#L443

Committed transactions are always inserted in-order (see here):
https://github.com/Expensify/Bedrock/blob/8d609f3005af4b5baded62e49e2d0747333f68f6/sqlitecluster/SQLite.cpp#L1223-L1238

All of these are sent over a single TCP connection, they can't arrive in a different order than they were sent.

Reply:
Got it. That explains everything, thanks!